aten/src/ATen/native/cpu/QuickPartition.cpp

```diff
@@ -0,0 +1,572 @@
+      left_in += nelts;
+      indices_left_in += nelts;
+    } else {
+      right_in -= nelts;
+      indices_right_in -= nelts;
+      Data3 = Vec256<T>::loadu(right_in);
+      Ind3 = Vec256<IndT>::loadu(indices_right_in);
+    }
+
+    // now write the output (plus some garbage) on both sides.
+    Ordered.store(right_out - nelts);
+    Ordered.store(left_out);
+    IndicesOrdered.store(indices_right_out - nelts);
+    IndicesOrdered.store(indices_left_out);
+
+    // update the output pointers depending on how much of the output above was real as opposed to garbage.
+    left_out += to_write_left;
+    right_out -= to_write_right;
+    indices_left_out += to_write_left;
+    indices_right_out -= to_write_right;
+
+    Data1 = Data2;
+    Data2 = Data3;
+    Ind1 = Ind2;
+    Ind2 = Ind3;
+  } while (right_in - left_in >= nelts);
+  // Main loop done.
+
+  // Now we have `2 * nelts` pieces of data still to use, in Data1 and Data2,
+  // and we also have some amount in [0, nelts) of data still to read,
+  // which we process in a normal scalar way.
+  while (left_in != right_in) {
+    T val;
+    IndT ind;
+    if (left_in <= left_out) {
+      val = *(--right_in);
+      ind = *(--indices_right_in);
+    } else {
+      val = *left_in++;
+      ind = *indices_left_in++;
+    }
+    bool valShouldGoLeft = largest ? !(pivot <= val) : (val < pivot);
+    if (valShouldGoLeft) {
+      *left_out++ = val;
+      *indices_left_out++ = ind;
+    } else {
+      *(--right_out) = val;
+      *(--indices_right_out) = ind;
+    }
+  }
+
+  // Now all data has been read, and all but 2 * nelts
+  // has been written. It is in Data1 and Data2. Write it out
+  // in the same way as in the main loop.
+  for (int i = 0; i < 2; ++i) {
+    Vec256<T> Comp = largest ? Data1 < Piv : Data1 > Piv;
+    int32_t const mask = Comp.msb_mask();
+    auto Perm = Vec256<IndT>::loadu(&(Table<float>::t[nelts*mask]));
+    auto Ordered
+      = permute(Data1, Perm);
+    auto IndicesOrdered
+      = permute(Ind1, Perm);
+
+    uint8_t to_write_right = __builtin_popcount(mask);
+    uint8_t to_write_left = nelts - to_write_right;
+
+    Ordered.store(right_out - nelts);
+    Ordered.store(left_out);
+    IndicesOrdered.store(indices_right_out - nelts);
+    IndicesOrdered.store(indices_left_out);
+
+    left_out += to_write_left;
+    right_out -= to_write_right;
+    indices_left_out += to_write_left;
+    indices_right_out -= to_write_right;
+
+    Data1 = Data2;
+    Ind1 = Ind2;
+  }
+  // We are done.
+  return left_out - begin;
+}
+
+
+int32_t at::native::vec_qs_partition_inplace(float *begin, float *end,
+    int32_t *indices,
+    float pivot,
+    bool largest) {
+  if (largest) {
+    return qs_partition_inplace<float, true>(begin, end, indices, pivot);
+  } else {
+    return qs_partition_inplace<float, false>(begin, end, indices, pivot);
+  }
+}
+#else
+
+namespace at {
+namespace native {
+namespace {
+// fallback to scalar when we don't have avx2
```

Comment:
Yeah, so this says it falls back when we don't have avx2, but you're doing this as a compile time test, which is not sufficient. It needs to be a runtime test.

Reply:
Grep for `REGISTER_DISPATCH` to see how to do this properly.

Reply:
Actually, I suspect you may have to do more work. The existing code for handling machines which can vectorize or not do this purely by recompiling Vec256 multiple times under different AVX settings. But in your case you need something a little different, you don't want to run the algorithm above *at all* when you don't have sufficient AVX support, you just want to do something completely different. I'm not exactly sure how to spell that presently. cc @colesbury  @cpuhrsch

Reply:
You can guard using __AVX__ macros

Reply:
Err, really? I thought `__AVX__` says whether or not the compiler has AVX support. But that's not what we need to control over here; we need to control whether or not the runtime has AVX support.